### PR TITLE
Add Assert.ContainsAll and Assert.DoesNotContainAll with structured assertion messages

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Adds `Assert.ContainsAll` and `Assert.DoesNotContainAll` (generic and non-generic, with and without an `IEqualityComparer`) on the `Assert` class, providing the same containment semantics as `CollectionAssert.IsSubsetOf` / `CollectionAssert.IsNotSubsetOf` but with the [RFC 012](../blob/main/docs/RFCs/012-Structured-Assertion-Messages.md) structured assertion message format and naming consistent with the existing `Assert.Contains(expected, collection)` API.

This is the first in a small series of PRs lifting `CollectionAssert.*` APIs onto `Assert.*` with the new message format.

## API surface

8 new public overloads:

```csharp
// Generic
Assert.ContainsAll<T>(IEnumerable<T>? expected, IEnumerable<T>? collection, string? message = "", ...)
Assert.ContainsAll<T>(IEnumerable<T>? expected, IEnumerable<T>? collection, IEqualityComparer<T>? comparer, string? message = "", ...)
Assert.DoesNotContainAll<T>(IEnumerable<T>? expected, IEnumerable<T>? collection, string? message = "", ...)
Assert.DoesNotContainAll<T>(IEnumerable<T>? expected, IEnumerable<T>? collection, IEqualityComparer<T>? comparer, string? message = "", ...)

// Non-generic
Assert.ContainsAll(IEnumerable? expected, IEnumerable? collection, string? message = "", ...)
Assert.ContainsAll(IEnumerable? expected, IEnumerable? collection, IEqualityComparer? comparer, string? message = "", ...)
Assert.DoesNotContainAll(IEnumerable? expected, IEnumerable? collection, string? message = "", ...)
Assert.DoesNotContainAll(IEnumerable? expected, IEnumerable? collection, IEqualityComparer? comparer, string? message = "", ...)
```

All overloads support `[CallerArgumentExpression]` capture for `expected` and `collection`.

The parameter naming mirrors the existing `Assert.Contains(expected, collection)` API: `expected` is the (collection of) item(s) we expect to find, `collection` is the container being searched.

## Behavior

- **Multiplicity-aware diffing**: each excess element in `expected` (with multiplicity) appears in the `missing:` evidence line — a true diff rather than a set-membership check.
- **Null handling**: nulls in `expected` are matched against nulls in `collection` (by count) without going through the user-supplied `IEqualityComparer`, matching the existing `CollectionAssert` semantics.
- **Single enumeration**: `expected` and `collection` are snapshotted to `List<T?>` once before counting, so lazy / single-pass enumerables behave deterministically and the failure path doesn't re-enumerate.
- **Comparer in evidence + call-site**: when a non-default comparer is supplied, RFC 012 requires the comparer's short type name in the evidence block and a `<comparer>` placeholder in the rendered call-site. Both are emitted.

## Example failure messages

```text
Assertion failed. Expected collection to contain all specified items.

missing:    [3]
expected:   [1, 2, 3]
collection: [1, 2]

Assert.ContainsAll(new[] { 1, 2, 3 }, new[] { 1, 2 })
```

With a non-default comparer:

```text
Assertion failed. Expected collection to contain all specified items.

missing:    ["C"]
expected:   ["A", "C"]
collection: ["a", "b"]
comparer:   OrdinalIgnoreCaseComparer

Assert.ContainsAll(new[] { "A", "C" }, new[] { "a", "b" }, <comparer>)
```

## Tests

35+ unit tests in `AssertTests.ContainsAll.cs` covering:

- Passing scenarios (all present, empty `expected`, multiplicity within `collection`).
- Failing scenarios with full structured-message verification.
- Multiplicity exceeding `collection`.
- Nulls in `expected`, nulls only in `collection`.
- Custom `IEqualityComparer<T>` (`StringComparer.OrdinalIgnoreCase`) + comparer-line + `<comparer>` placeholder in call-site.
- Non-generic `IEnumerable` variants (with and without comparer).
- All `[NotNull]` parameter contracts (`expected`, `collection`, `comparer`) for all 8 overloads.
- Empty-collection edge cases for `DoesNotContainAll` (both empty / empty `expected` both correctly fail).
- `AssertFailedException.ExpectedText` / `ActualText` properties are populated correctly.

All tests pass; full repo `Build.cmd` succeeds with 0 warnings, 0 errors.

## Notes

- 2 new resource entries in `FrameworkMessages.resx`; `.xlf` files were regenerated by the build (not hand-edited).
- 8 new entries added to `PublicAPI.Unshipped.txt`.
- No existing public API was modified.

### CODE REVIEW

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+    {
+        string? callSite = FormatCallSiteExpression(assertionMethodName, subsetExpression, supersetExpression, "<subset>", "<superset>");
+        if (callSite is null || !hasComparer)
+        {
+            return callSite;
+        }
+
+        // FormatCallSiteExpression has no overload accepting a third argument expression; insert
+        // the <comparer> placeholder so the rendered call-site reflects the overload that was actually invoked.
+        return string.Concat(callSite.Substring(0, callSite.Length - 1), ", <comparer>)");
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: **[MODERATE]** This string manipulation is fragile — it assumes `FormatCallSiteExpression` always returns a string that ends exactly with `)`. If the format ever changes (e.g., trailing whitespace, different terminator), the injected `, <comparer>)` produces a malformed call-site string.

Consider either:
1. Adding a `FormatCallSiteExpression` overload that accepts a third argument expression, making the insertion explicit and safe.
2. Using `callSite[..^1]` (range syntax) here at minimum for idiomatic C#, and adding a defensive assertion/check that `callSite.EndsWith(')')` before manipulating it.

```csharp
// Defensive guard:
Debug.Assert(callSite.EndsWith(')'), "FormatCallSiteExpression output format changed");
return callSite[..^1] + ", <comparer>)";
```

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+    }
+
+    private sealed class NonGenericEqualityComparerAdapter : IEqualityComparer<object?>
+    {
+        private readonly IEqualityComparer _comparer;
+
+        public NonGenericEqualityComparerAdapter(IEqualityComparer comparer)
+            => _comparer = comparer;
```
Comment (Evangelink) [inline src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: **[NIT]** `public new bool Equals` shadows `object.Equals(object?)` by name only (they differ in arity, and `object.Equals(object?, object?)` is static). The `new` is required to suppress the warning, which is correct here. Just noting for reviewers that this is intentional and not a hidden-method smell.

**Review by Evangelink — COMMENTED** [review]
## PR #8234 — `Assert.IsSubsetOf` / `Assert.IsNotSubsetOf`

Good implementation. Multiplicity-aware diffing, single-enumeration snapshots, null-safe counting, proper structured messages, correct `PublicAPI.Unshipped.txt` entries, auto-generated XLF. The test coverage is solid (35 tests covering generic/non-generic, comparer, null inputs, edge cases, and `ExpectedText`/`ActualText` population).

---

### Verdict Table

| # | Dimension | Status | Notes |
|---|-----------|--------|-------|
| 1 | Algorithmic Correctness | ✅ LGTM | Empty subset, null handling, multiplicity all correct |
| 2 | Threading & Concurrency | ✅ N/A | No shared mutable state |
| 3 | Security | ✅ N/A | No IPC/deserialization |
| 4 | Public API & Binary Compat | ✅ LGTM | 8 entries in `PublicAPI.Unshipped.txt`, no `init` accessors, `sealed` class |
| 5 | Performance & Allocations | ✅ LGTM | Single enumeration, `List<T?>` snapshot reuse |
| 6 | Cross-TFM Compatibility | ✅ LGTM | No TFM-gated API used |
| 7 | Resource & IDisposable | ✅ N/A | No disposables |
| 8 | Defensive Coding | ✅ LGTM | All params null-checked before use |
| 9 | Localization & Resources | ✅ LGTM | `.resx` updated, XLF auto-generated |
| 10 | Test Isolation | ✅ LGTM | No shared static mutable state |
| 11 | Assertion Quality | ✅ LGTM | Uses `TestFramework.ForTestingMSTest` + `AwesomeAssertions` |
| 12 | Flakiness | ✅ N/A | No sleeps or order-dependent collection asserts |
| 13 | Test Completeness | ✅ LGTM | All 8 overloads exercised; null/edge cases covered |
| 14 | Data-Driven Coverage | ✅ N/A | Direct tests sufficient |
| 15 | Code Structure | ✅ LGTM | Pattern matching, `is null`/`is not null` throughout |
| 16 | Naming & Conventions | ✅ LGTM | File-scoped namespace, consistent with codebase |
| 17 | Documentation | ✅ LGTM | Full XML docs on all 8 public overloads |
| 18 | Analyzer Quality | ✅ N/A | No analyzer changes |
| 19 | IPC Wire Compat | ✅ N/A | No serialization |
| 20 | Build Infrastructure | ✅ N/A | No `eng/` changes |
| 21 | Scope & PR Discipline | ✅ LGTM | Single focused concern |

---

### Key Findings

**[MODERATE]** `BuildCallSiteWithComparer` (line 454) performs string surgery by stripping the trailing `)` from `FormatCallSiteExpression`'s output and appending `, <comparer>)`. This silently produces a malformed call-site string if the format ever changes. The best fix is a new `FormatCallSiteExpression` overload that accepts a third argument expression; a defensive `Debug.Assert(callSite.EndsWith(')'))` is a reasonable short-term guard. See inline comment.

**[NIT]** `public new bool Equals` in `NonGenericEqualityComparerAdapter` is correct and intentional (implements `IEqualityComparer<object?>`), but the `new` keyword warrants a brief comment so future readers don't question it. See inline comment.

**[NIT — test gap]** The non-generic + comparer *happy-path* cases (`IsSubsetOf_NonGeneric_WithComparer_AllPresent_ShouldPass` and `IsNotSubsetOf_NonGeneric_WithComparer_MissingElement_ShouldPass`) are absent. Not blocking, but the symmetry is worth closing.

> Generated by [Expert Code Review (on open)](https://github.com/microsoft/testfx/actions/runs/25880871221/agentic_workflow) for issue #8234 · ● 10.7M

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+        string? callSite = FormatCallSiteExpression(assertionMethodName, subsetExpression, supersetExpression, "<subset>", "<superset>");
+        if (callSite is null || !hasComparer)
+        {
+            return callSite;
+        }
+
+        // FormatCallSiteExpression has no overload accepting a third argument expression; insert
+        // the <comparer> placeholder so the rendered call-site reflects the overload that was actually invoked.
+        return string.Concat(callSite.Substring(0, callSite.Length - 1), ", <comparer>)");
+    }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: Addressed in 02ad596 — added a 3-expression overload of `FormatCallSiteExpression` and refactored `BuildCallSiteWithComparer` to use it, eliminating the substring surgery.

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+
+    private sealed class NonGenericEqualityComparerAdapter : IEqualityComparer<object?>
+    {
+        private readonly IEqualityComparer _comparer;
+
+        public NonGenericEqualityComparerAdapter(IEqualityComparer comparer)
+            => _comparer = comparer;
+
+        public new bool Equals(object? x, object? y) => _comparer.Equals(x, y);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: Addressed in 02ad596 — the `new` is intentional (it suppresses CS0108 because `object` exposes a *static* `Equals(object?, object?)`; this instance method satisfies `IEqualityComparer<object?>.Equals` and intentionally hides the static one by name only). I added a code comment explaining that so future readers don't question it.

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+            if (entry.Value > supersetCount)
+            {
+                missing ??= [];
+                int excess = entry.Value - supersetCount;
+                for (int i = 0; i < excess; i++)
+                {
+                    missing.Add(entry.Key);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: Addressed in 02ad596 (same fix as the sibling comment) — `TryFindMissingElements` now walks `subset` in source order, so `missing` is in true first-seen order without relying on `Dictionary` enumeration order.

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+    /// <summary>
+    /// Determines whether <paramref name="subset"/> contains any element not present
+    /// (with sufficient multiplicity) in <paramref name="superset"/>.
+    /// </summary>
+    /// <returns>
+    /// <see langword="true"/> if at least one element is missing — in which case <paramref name="missing"/>
+    /// holds the excess elements (in their first-seen order) — and <see langword="false"/> when every
+    /// element of <paramref name="subset"/> is matched in <paramref name="superset"/>.
+    /// </returns>
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: Addressed in 02ad596. `TryFindMissingElements` now walks the `subset` in source order and decrements the per-key superset quota on the fly, so `missing` is in true first-seen positional order (with multiplicity preserved) and no longer depends on `Dictionary` enumeration order.

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs (hunk 1 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: Addressed in 02ad596 by going with option 1 — added a 3-expression overload of `FormatCallSiteExpression` and refactored `BuildCallSiteWithComparer` to use it. No more string surgery.

File: src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs (hunk 2 above)
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.IsSubsetOf.cs]: Addressed in 02ad596 — added an inline code comment on `NonGenericEqualityComparerAdapter.Equals` explaining the intent of the `new` modifier.

File: test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.ContainsAll.cs [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+        ex.ActualText.Should().Be("[1, 2]");
+    }
+
+    private sealed class CaseInsensitiveStringComparer : IEqualityComparer<string>, IEqualityComparer
+    {
+        public bool Equals(string? x, string? y) => StringComparer.OrdinalIgnoreCase.Equals(x, y);
+
+        public int GetHashCode(string obj) => StringComparer.OrdinalIgnoreCase.GetHashCode(obj);
+
+        bool IEqualityComparer.Equals(object? x, object? y) => StringComparer.OrdinalIgnoreCase.Equals(x as string, y as string);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply test/UnitTests/TestFramework.UnitTests/Assertions/AssertTests.ContainsAll.cs]: Fixed in 1f381d2c8b97a9d7d872b3a37e370c80c73b3edb: the non-generic comparer now performs explicit null/type checks, only delegates to the string comparer for string inputs, uses RuntimeHelpers.GetHashCode for non-string values, and adds regression coverage for both behaviors.

File: src/TestFramework/TestFramework/Assertions/Assert.ContainsAll.cs [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+        structured.WithEvidence(evidence);
+        structured.WithExpectedAndActual(expectedText: null, actualText: collectionText);
+        structured.WithCallSiteExpression(BuildCallSiteWithComparer("Assert.DoesNotContainAll", expectedExpression, collectionExpression, comparerName is not null));
+
+        ReportAssertFailed(structured);
+    }
+
+    private static string? BuildCallSiteWithComparer(string assertionMethodName, string expectedExpression, string collectionExpression, bool hasComparer)
+        => hasComparer
+            ? FormatCallSiteExpression(assertionMethodName, expectedExpression, collectionExpression, expression3: string.Empty, "<expected>", "<collection>", "<comparer>")
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/TestFramework/TestFramework/Assertions/Assert.ContainsAll.cs]: Good catch. Fixed in d007f5638 by passing `"<comparer>"` (instead of `string.Empty`) as `expression3` in `BuildCallSiteWithComparer`. Because the comparer parameter doesn't carry a `[CallerArgumentExpression]`, the third slot was always empty, which combined with the all-empty short-circuit in `FormatCallSiteExpression` could suppress the call site entirely when callers (e.g. wrappers or non-C# languages) don't propagate caller expressions. With this change the call site now always renders, and degrades to `Assert.ContainsAll(<expected>, <collection>, <comparer>)` (and likewise for `DoesNotContainAll`) when no expressions are available. Added two unit tests in `AssertTests.ContainsAll.cs` covering the missing-caller-info case for both APIs.

**Comment by Evangelink** [comment]
@copilot resolve the merge conflicts in this pull request
